### PR TITLE
br: push down checkpoint record

### LINKED ISSUE #43632: checkpoint for snapshot restore should be more accurate
## Enhancement
Currently the checkpoint for snapshot restore records the import done ranges only when a batch finished, which contains 128 ranges in default. What's more, it is fair enough for each batch to seize `worker` in the workerpool. Therefore, most batch will be completed in the far and concentrated future.

So we need to records the ranges once the range has been import done.

What's more, the current value of checkpoint data is the `merged range's start key`. There is a risk of changing the merged key if the config `mergeRegionSize` and `mergeRegionCount` is changed.

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve?

Issue Number: close #43632 

Problem Summary:
checkpoint for snapshot restore should be more accurate.
### What is changed and how it works?
records the ranges once the range has been import done.

Besides, the changed checkpoint data for snapshot restore now is :
```
<table-id, start-key>
```
instead of the original one :
```
<table-id, (start-key, end-key)>
```
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/checkpoint/checkpoint_test.go [hunk 1]
```diff
@@ -276,29 +264,23 @@ func TestCheckpointRestoreRunner(t *testing.T) {
...
-			d, ok = data2[string(resp.StartKey)]
+			d, ok = data2[resp.RangeKey]
 			require.Equal(t, tableID, int64(2))
 			require.True(t, ok)
 		} else {
 			require.Equal(t, tableID, int64(1))
 		}
-		require.Equal(t, d.StartKey, string(resp.StartKey))
-		require.Equal(t, d.EndKey, string(resp.EndKey))
+		require.Equal(t, d.RangeKey, string(resp.RangeKey))
```
Comment (YuJuncen) [inline br/pkg/checkpoint/checkpoint_test.go:283]: br/pkg/checkpoint/checkpoint_test.go:283:32: unnecessary conversion (unconvert)

**Review by YuJuncen — APPROVED** [review]
It seems this PR also make the checkpoint key from `(StartKey, EndKey)` tuple become solely the `StartKey`, could you add some comment about this change in this PR?
Generally LGTM.

File: br/pkg/restore/client.go [hunk 2]
```diff
@@ -1258,27 +1258,36 @@ func (rc *Client) setSpeedLimit(ctx context.Context, rateLimit uint64) error {
...
-	idx1 := strings.LastIndex(f1, "_")
-	idx2 := strings.LastIndex(f2, "_")
-
-	if idx1 < 0 || idx2 < 0 {
-		panic(fmt.Sprintf("invalid backup data file name: '%s', '%s'", f1, f2))
-	}
+	idx1 := getFileRangeKeyIndex(f1)
+	idx2 := getFileRangeKeyIndex(f2)
 
 	return f1[:idx1] == f2[:idx2]
```
Comment (3pointer) [inline br/pkg/restore/client.go:1284]:
```suggestion
	return getFileRangeKey(f1) == getFileRangeKey(f2)
```

File: br/pkg/restore/client.go [hunk 3]
```diff
@@ -1258,27 +1258,36 @@ func (rc *Client) setSpeedLimit(ctx context.Context, rateLimit uint64) error {
...
 }
 
-func drainFilesByRange(files []*backuppb.File, supportMulti bool) ([]*backuppb.File, []*backuppb.File) {
+func drainFilesByRange(files []*backuppb.File) ([]*backuppb.File, []*backuppb.File) {
 	if len(files) == 0 {
 		return nil, nil
 	}
-	if !supportMulti {
```
Comment (3pointer) [inline br/pkg/restore/client.go:1279]: why remove it ?

File: br/pkg/restore/client.go [hunk 4]
```diff
@@ -1379,29 +1403,54 @@ func (rc *Client) RestoreSSTFiles(
...
+		for rangeFiles, leftFiles = drainFilesByRange(files); len(rangeFiles) != 0; rangeFiles, leftFiles = drainFilesByRange(leftFiles) {
+			filesReplica := rangeFiles
+			if ectx.Err() != nil {
+				log.Warn("Restoring encountered error and already stopped, give up remained files.",
+					zap.Int("remained", len(leftFiles)),
+					logutil.ShortError(ectx.Err()))
+				// We will fetch the error from the errgroup then (If there were).
+				// Also note if the parent context has been canceled or something,
+				// breaking here directly is also a reasonable behavior.
+				break
```
Comment (3pointer) [inline br/pkg/restore/client.go:1423]: need break `for _, tableIDWithFile := range tableIDWithFiles ` loop

File: br/pkg/restore/client.go [hunk 5]
```diff
@@ -1379,29 +1393,54 @@ func (rc *Client) RestoreSSTFiles(
...
+						return errors.Trace(importErr)
+					}
+				}
+
+				// the data of this range has been import done
+				if runner != nil && len(filesReplica) > 0 {
+					rangeKey := getFileRangeKey(filesReplica[0].Name)
+					// The checkpoint range shows this ranges of kvs has been restored into
+					// the table corresponding to the table-id.
+					if err := checkpoint.AppendRangesForRestore(ectx, runner, tableID, rangeKey); err != nil {
```
Comment (3pointer) [inline br/pkg/restore/client.go:1437]: does checkpoint.AppendRangesForRestore thread-safe?

**Review by 3pointer — APPROVED** [review]

**Pull request merged**